numba/typing/setdecl.py

```diff
@@ -0,0 +1,182 @@
+                dtype = iterable.iterator_type.yield_type
+                if isinstance(dtype, types.Hashable):
+                    return signature(types.Set(dtype), iterable)
+        else:
+            # set()
+            return signature(types.Set(types.undefined))
+
+
+@infer_getattr
+class SetAttribute(AttributeTemplate):
+    key = types.Set
+
+    @bound_function("set.add")
+    def resolve_add(self, set, args, kws):
+        item, = args
+        assert not kws
+        unified = self.context.unify_pairs(set.dtype, item)
+        if unified is not None:
+            sig = signature(types.none, unified)
+            sig.recvr = set.copy(dtype=unified)
+            return sig
+
+    @bound_function("set.clear")
+    def resolve_clear(self, set, args, kws):
+        assert not kws
+        if not args:
+            return signature(types.none)
+
+    @bound_function("set.copy")
+    def resolve_copy(self, set, args, kws):
+        assert not kws
+        if not args:
+            return signature(set)
+
+    @bound_function("set.discard")
+    def resolve_discard(self, set, args, kws):
+        item, = args
+        assert not kws
+        return signature(types.none, set.dtype)
+
+    @bound_function("set.pop")
+    def resolve_pop(self, set, args, kws):
+        assert not kws
+        if not args:
+            return signature(set.dtype)
+
+    @bound_function("set.remove")
+    def resolve_remove(self, set, args, kws):
+        item, = args
+        assert not kws
+        return signature(types.none, set.dtype)
+
+    @bound_function("set.update")
+    def resolve_update(self, set, args, kws):
+        iterable, = args
+        assert not kws
+        if not isinstance(iterable, types.IterableType):
+            return
+
+        dtype = iterable.iterator_type.yield_type
+        unified = self.context.unify_pairs(set.dtype, dtype)
+        if unified is not None:
+            sig = signature(types.none, iterable)
+            sig.recvr = set.copy(dtype=unified)
+            return sig
+
+    def _resolve_xxx_update(self, set, args, kws):
+        assert not kws
+        iterable, = args
+        # Set arguments only supported for now
+        if iterable == set:
+            return signature(types.none, iterable)
+
+    @bound_function("set.difference_update")
+    def resolve_difference_update(self, set, args, kws):
+        return self._resolve_xxx_update(set, args, kws)
+
+    @bound_function("set.intersection_update")
+    def resolve_intersection_update(self, set, args, kws):
+        return self._resolve_xxx_update(set, args, kws)
+
+    @bound_function("set.symmetric_difference_update")
+    def resolve_symmetric_difference_update(self, set, args, kws):
+        return self._resolve_xxx_update(set, args, kws)
+
+    def _resolve_operator(self, set, args, kws):
+        assert not kws
+        iterable, = args
+        # Set arguments only supported for now
+        if iterable == set:
```

Comment:
This prevents operations on reflected and non-reflected sets together; e.g. ``non_reflected_set.union(reflected_set)`` and vice versa.